vocode/streaming/agent/state_agent.py

```diff
@@ -278,80 +279,110 @@ async def choose_block(self, state=None, choose_from=None):
+            "condition": "[insert the name of the condition that applies]",
         }
-        choices = "\n".join(
-            [
-                f"- '{c.lower()}'"
-                for c in state["condition"]["conditionToStateLabel"].keys()
-            ]
-        )
-        # self.logger.info(f"Choosing condition from:\n{choices}")
-        # check if there is more than one condition
-        prompt = (
+        prompt = (  # i was gonna move a negation of the current state's relevance to choces
+            f"The bot helping with: '{self.current_state['id'].split('::')[0]}'\n"
             f"Bot's last statement: '{last_bot_message}'\n"
             f"User's response: '{last_user_message}'\n"
-            "Identify the most fitting condition from the list below:\n"
-            f"{choices}\n"
-            "Instructions:\n"
-            "- Select the condition that applies.\n"
-            "- Provide the exact name of the condition.\n"
-            "- If no conditions apply, type 'none'.\n"
-            "- If the user asked a question, type 'question'."
+            f"{'Last tool output: ' + last_tool_output if last_tool_output else ''}\n\n"
+            "Identify the name of the most fitting condition from the list below:\n"
+            f"{choices}\n\n"
+            "Always return a condition from the above list. Return the exact name of the condition that best applies."
         )
         self.logger.info(f"AI prompt constructed: {prompt}")
-        if len(state["condition"]["conditionToStateLabel"]) > 1:
-            response = await self.call_ai(
-                prompt,
-                tool,
-            )
-        else:  # with just a single condition, change the prompt
-            single_condition = list(state["condition"]["conditionToStateLabel"].keys())[
-                0
-            ]
-            response = await self.call_ai(
-                f"Your previous statement was: '{last_bot_message}', to which the user replied: '{last_user_message}'.\n\nBased on your instructions and the context, assess whether the condition named '{single_condition}' currently applies.\n\n- If it is valid, provide the exact name of the condition.\n- If it does not apply, provide 'none'.\n- In the event that the user's reply is a question, provide 'question'.",
-                tool,
-            )
+        response = await self.call_ai(
+            prompt,
+            tool,
+        )
         self.logger.info(f"Chose condition: {response}")
         response = response[response.find("{") : response.rfind("}") + 1]
         try:
             response = eval(response)
             condition = response["condition"]
-            if condition.lower() == "question":
-                move_on = await self.maybe_respond_to_user(
-                    last_bot_message, last_user_message
-                )
-                if not move_on:
-                    return
```

Comment:
I think this is where the bugginess is coming from. State machine traversal is recursive (although I'm thinking of changing that, lmk if you have a pref), and this code path doesn't call `handle_state()`, so I think the conversation would just end here

It looks like `maybe_respond_to_user` returns `False` when it wants to respond to user input? In that case I think I'll return a `resume` function that calls `maybe_respond_to_user`

note to self to remove the `choose_block` invocation in `maybe_respond_to_user`